File: android/app/src/test/kotlin/com/endurain/endurain/activity/MovementAutoPauseDetectorTest.kt
package com.endurain.endurain.activity

import org.junit.Assert.assertEquals
import org.junit.Test

/**
 * JVM unit tests for [MovementAutoPauseDetector].
 *
 * Mirrors the Dart test suite for
 * `lib/features/activity/services/movement_auto_pause_detector.dart`. A
 * regression here would either fail to auto-pause a stopped recording or
 * (worse) auto-resume when it should not, so every branch of the reliable-
 * movement heuristic and the pause/resume hysteresis is locked down.
 */
class MovementAutoPauseDetectorTest {

    private val baseMillis = 1_700_000_000_000L

    private fun sample(
        second: Int,
        lat: Double = 38.7,
        lon: Double = -9.1,
        speed: Double? = null,
        accuracy: Double? = 5.0,
    ): MovementSample {
        return MovementSample(
            timestampMillis = baseMillis + second * 1_000L,
            latitude = lat,
            longitude = lon,
            speedMetersPerSecond = speed,
            horizontalAccuracyMeters = accuracy,
        )
    }

    // ── onActivePoint: auto-pause ──────────────────────────────────────────

    @Test
    fun disabledConfigNeverAutoPauses() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(enabled = false, pauseDelayMillis = 1_000L),
        )
        detector.reset(baseMillis)

        for (second in 0..10) {
            val transition = detector.onActivePoint(sample(second, speed = 0.0))
            assertEquals(MovementAutoPauseTransition.NONE, transition)
        }
    }

    @Test
    fun autoPausesAfterStillnessPersistsForTheConfiguredDelay() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(enabled = true, pauseDelayMillis = 5_000L),
        )
        detector.reset(baseMillis)

        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onActivePoint(sample(1, speed = 0.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onActivePoint(sample(4, speed = 0.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.AUTO_PAUSE,
            detector.onActivePoint(sample(5, speed = 0.0)),
        )
    }

    @Test
    fun reportedSpeedAboveThresholdResetsTheStillnessTimer() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(enabled = true, pauseDelayMillis = 5_000L),
        )
        detector.reset(baseMillis)

        detector.onActivePoint(sample(4, speed = 0.0))
        // A moving sample just before the delay would have elapsed resets the
        // clock, so five more seconds of stillness are required.
        detector.onActivePoint(sample(4, speed = 2.0))
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onActivePoint(sample(8, speed = 0.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.AUTO_PAUSE,
            detector.onActivePoint(sample(9, speed = 0.0)),
        )
    }

    @Test
    fun unreliableAccuracyIsNeverCountedAsMovement() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(
                enabled = true,
                pauseDelayMillis = 5_000L,
                maxAccuracyMeters = 30.0,
            ),
        )
        detector.reset(baseMillis)

        // A fast-looking but inaccurate fix must not reset the stillness timer.
        detector.onActivePoint(sample(1, speed = 10.0, accuracy = 500.0))
        assertEquals(
            MovementAutoPauseTransition.AUTO_PAUSE,
            detector.onActivePoint(sample(5, speed = 0.0)),
        )
    }

    @Test
    fun displacementDerivedSpeedIsUsedWhenReportedSpeedIsAbsent() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(
                enabled = true,
                pauseDelayMillis = 5_000L,
                movingSpeedThresholdMetersPerSecond = 0.6,
            ),
        )
        detector.reset(baseMillis)

        // ~0.0009 degrees of latitude is roughly 100 meters; moving that far in
        // one second is far above the 0.6 m/s threshold, so the timer resets.
        detector.onActivePoint(sample(0, lat = 38.7, speed = null))
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onActivePoint(sample(1, lat = 38.7009, speed = null)),
        )
        // Two stationary (displacement ~0) fixes after that should reach the
        // configured delay and auto-pause.
        detector.onActivePoint(sample(2, lat = 38.7009, speed = null))
        assertEquals(
            MovementAutoPauseTransition.AUTO_PAUSE,
            detector.onActivePoint(sample(6, lat = 38.7009, speed = null)),
        )
    }

    // ── onAutoPausedPoint: auto-resume ──────────────────────────────────────

    @Test
    fun autoResumesOnlyAfterEnoughConsecutiveMovingSamples() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(
                enabled = true,
                consecutiveMovingSamplesToResume = 3,
            ),
        )
        detector.reset(baseMillis)

        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onAutoPausedPoint(sample(1, speed = 2.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onAutoPausedPoint(sample(2, speed = 2.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.AUTO_RESUME,
            detector.onAutoPausedPoint(sample(3, speed = 2.0)),
        )
    }

    @Test
    fun aSingleStillSampleResetsTheConsecutiveMovingCounter() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(
                enabled = true,
                consecutiveMovingSamplesToResume = 3,
            ),
        )
        detector.reset(baseMillis)

        detector.onAutoPausedPoint(sample(1, speed = 2.0))
        detector.onAutoPausedPoint(sample(2, speed = 2.0))
        // Movement noise briefly drops out; hysteresis must not resume yet.
        detector.onAutoPausedPoint(sample(3, speed = 0.0))
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onAutoPausedPoint(sample(4, speed = 2.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onAutoPausedPoint(sample(5, speed = 2.0)),
        )
        assertEquals(
            MovementAutoPauseTransition.AUTO_RESUME,
            detector.onAutoPausedPoint(sample(6, speed = 2.0)),
        )
    }

    @Test
    fun resetClearsStillnessAndConsecutiveMovementHistory() {
        val detector = MovementAutoPauseDetector(
            MovementAutoPauseConfig(enabled = true, pauseDelayMillis = 1_000L),
        )
        detector.reset(baseMillis)
        detector.onActivePoint(sample(5, speed = 0.0))

        // A manual pause/resume resets tracking; the next active point must
        // not immediately auto-pause using stale history.
        detector.reset(baseMillis + 100_000L)
        assertEquals(
            MovementAutoPauseTransition.NONE,
            detector.onActivePoint(sample(100, speed = 0.0)),
        )
    }

    // ── haversineMeters ──────────────────────────────────────────────────────

    @Test
    fun haversineMetersIsZeroForIdenticalCoordinates() {
        assertEquals(
            0.0,
            MovementAutoPauseDetector.haversineMeters(38.7, -9.1, 38.7, -9.1),
            0.0001,
        )
    }

    @Test
    fun haversineMetersApproximatesKnownShortDistance() {
        // One arc-minute of latitude is ~1852 meters (a nautical mile).
        val meters = MovementAutoPauseDetector.haversineMeters(0.0, 0.0, 1.0 / 60.0, 0.0)
        assertEquals(1852.0, meters, 5.0)
    }
}

File: test/features/activity/services/movement_auto_pause_detector_test.dart
import 'package:endurain/features/activity/models/recorded_activity_point.dart';
import 'package:endurain/features/activity/services/movement_auto_pause_detector.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  group('MovementAutoPauseDetector', () {
    final base = DateTime.utc(2026, 6, 3, 9);

    RecordedActivityPoint point({
      double latitude = 38.7,
      double longitude = -9.1,
      int second = 0,
      double? speed,
      double? accuracy = 5,
    }) {
      return RecordedActivityPoint(
        timestamp: base.add(Duration(seconds: second)),
        latitude: latitude,
        longitude: longitude,
        segmentIndex: 0,
        speedMetersPerSecond: speed,
        horizontalAccuracyMeters: accuracy,
      );
    }

    group('onActivePoint (auto-pause)', () {
      test('disabled config never auto-pauses', () {
        final detector = MovementAutoPauseDetector(
          config: const MovementAutoPauseConfig(
            enabled: false,
            pauseDelay: Duration(seconds: 1),
          ),
        )..reset(movementAt: base);

        for (var second = 0; second <= 10; second++) {
          final transition = detector.onActivePoint(
            point(second: second, speed: 0),
          );
          expect(transition, MovementAutoPauseTransition.none);
        }
      });

      test('auto-pauses once stillness persists for the configured delay', () {
        final detector = MovementAutoPauseDetector(
          config: const MovementAutoPauseConfig(
            enabled: true,
            pauseDelay: Duration(seconds: 5),
          ),
        )..reset(movementAt: base);

        expect(
          detector.onActivePoint(point(second: 1, speed: 0)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onActivePoint(point(second: 4, speed: 0)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onActivePoint(point(second: 5, speed: 0)),
          MovementAutoPauseTransition.autoPause,
        );
      });

      test('reported speed above the threshold resets the stillness timer', () {
        final detector = MovementAutoPauseDetector(
          config: const MovementAutoPauseConfig(
            enabled: true,
            pauseDelay: Duration(seconds: 5),
          ),
        )..reset(movementAt: base);

        detector.onActivePoint(point(second: 3, speed: 0));
        detector.onActivePoint(point(second: 4, speed: 2));
        expect(
          detector.onActivePoint(point(second: 8, speed: 0)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onActivePoint(point(second: 9, speed: 0)),
          MovementAutoPauseTransition.autoPause,
        );
      });

      test('unreliable accuracy is never counted as movement', () {
        final detector = MovementAutoPauseDetector(
          config: const MovementAutoPauseConfig(
            enabled: true,
            pauseDelay: Duration(seconds: 5),
            maxAccuracyMeters: 30,
          ),
        )..reset(movementAt: base);

        // A fast-looking but inaccurate fix must not reset the stillness timer.
        detector.onActivePoint(point(second: 1, speed: 10, accuracy: 500));
        expect(
          detector.onActivePoint(point(second: 5, speed: 0)),
          MovementAutoPauseTransition.autoPause,
        );
      });

      test(
        'displacement-derived speed is used when reported speed is absent',
        () {
          final detector = MovementAutoPauseDetector(
            config: const MovementAutoPauseConfig(
              enabled: true,
              pauseDelay: Duration(seconds: 5),
              movingSpeedThresholdMetersPerSecond: 0.6,
            ),
          )..reset(movementAt: base);

          // ~0.0009 degrees of latitude is roughly 100 meters; moving that far
          // in one second is far above the 0.6 m/s threshold, so the timer
          // resets.
          detector.onActivePoint(point(second: 0, latitude: 38.7));
          expect(
            detector.onActivePoint(point(second: 1, latitude: 38.7009)),
            MovementAutoPauseTransition.none,
          );
          // Two stationary (displacement ~0) fixes after that should reach the
          // configured delay and auto-pause.
          detector.onActivePoint(point(second: 2, latitude: 38.7009));
          expect(
            detector.onActivePoint(point(second: 6, latitude: 38.7009)),
            MovementAutoPauseTransition.autoPause,
          );
        },
      );
    });

    group('onAutoPausedPoint (auto-resume)', () {
      test('auto-resumes only after enough consecutive moving samples', () {
        final detector = MovementAutoPauseDetector(
          config: const MovementAutoPauseConfig(
            enabled: true,
            consecutiveMovingSamplesToResume: 3,
          ),
        )..reset(movementAt: base);

        expect(
          detector.onAutoPausedPoint(point(second: 1, speed: 2)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onAutoPausedPoint(point(second: 2, speed: 2)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onAutoPausedPoint(point(second: 3, speed: 2)),
          MovementAutoPauseTransition.autoResume,
        );
      });

      test('a single still sample resets the consecutive-moving counter', () {
        final detector = MovementAutoPauseDetector(
          config: const MovementAutoPauseConfig(
            enabled: true,
            consecutiveMovingSamplesToResume: 3,
          ),
        )..reset(movementAt: base);

        detector.onAutoPausedPoint(point(second: 1, speed: 2));
        detector.onAutoPausedPoint(point(second: 2, speed: 2));
        // Movement noise briefly drops out; hysteresis must not resume yet.
        detector.onAutoPausedPoint(point(second: 3, speed: 0));
        expect(
          detector.onAutoPausedPoint(point(second: 4, speed: 2)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onAutoPausedPoint(point(second: 5, speed: 2)),
          MovementAutoPauseTransition.none,
        );
        expect(
          detector.onAutoPausedPoint(point(second: 6, speed: 2)),
          MovementAutoPauseTransition.autoResume,
        );
      });
    });

    test('reset clears stillness and consecutive-movement history', () {
      final detector = MovementAutoPauseDetector(
        config: const MovementAutoPauseConfig(
          enabled: true,
          pauseDelay: Duration(seconds: 1),
        ),
      )..reset(movementAt: base);
      detector.onActivePoint(point(second: 5, speed: 0));

      // A manual pause/resume resets tracking; the next active point must not
      // immediately auto-pause using stale history.
      detector.reset(movementAt: base.add(const Duration(seconds: 100)));
      expect(
        detector.onActivePoint(point(second: 100, speed: 0)),
        MovementAutoPauseTransition.none,
      );
    });
  });
}
